iib/web/api_v1.py

```diff
@@ -37,3 +38,94 @@ def add_bundles():
+    """
+    allowed_users = flask.current_app.config['IIB_WORKER_USERNAMES']
+    # current_user.is_authenticated is only ever False when auth is disabled
+    if current_user.is_authenticated and current_user.username not in allowed_users:
+        raise Unauthorized('This API endpoint is restricted to IIB workers')
+
+    payload = flask.request.get_json()
+    if not isinstance(payload, dict):
+        raise ValidationError('The input data must be a JSON object')
+
+    if not payload:
+        raise ValidationError('At least one key must be specified to update the request')
+
+    valid_keys = {
+        'arches', 'bundles', 'binary_image_resolved', 'from_index_resolved', 'index_image',
+        'state', 'state_reason'
+    }
+    invalid_keys = payload.keys() - valid_keys
+    if invalid_keys:
+        raise ValidationError(
+            'The following keys are not allowed: {}'.format(', '.join(invalid_keys)))
+
+    for key, value in payload.items():
+        if key in ('bundles', 'arches'):
+            if (
+                    not isinstance(value, list) or
+                    any(
+                        not param_value or not isinstance(param_value, str)
+                        for param_value in payload.get(key)
+                    )
+            ):
+                raise ValidationError(f'The value for "{key}" must be a non-empty string array.')
+
+        if key in ('binary_image_resolved', 'from_index_resolved', 'index_image'):
+            if not value or not isinstance(value, str):
+                raise ValidationError(f'The value for "{key}" must be a non-empty string')
+
+    if 'state' in payload and 'state_reason' not in payload:
+        raise ValidationError('The "state_reason" key is required when "state" is supplied')
+    elif 'state_reason' in payload and 'state' not in payload:
+        raise ValidationError('The "state" key is required when "state_reason" is supplied')
+
+    request = Request.query.get_or_404(request_id)
+    if 'state' in payload and 'state_reason' in payload:
+        new_state = payload['state']
+        new_state_reason = payload['state_reason']
+        # This is to protect against a Celery task getting executed twice and setting the
+        # state each time
+        if request.state.state == new_state and request.state.state_reason == new_state_reason:
+            flask.current_app.logger.info('Not adding a new state since it matches the last state')
+        else:
+            request.add_state(new_state, new_state_reason)
+
+    if 'bundles' in payload:
+        for bundle in payload.get('bundles', []):
+            image_object = Image.get_or_create(bundle)
+            if image_object not in request.bundles:
+                request.bundles.append(image_object)
+
+    for key in ('binary_image_resolved', 'from_index_resolved', 'index_image'):
+        if key not in payload:
+            continue
+        if key == 'index_image' and 'arches' in payload:
```

Comment:
The `arches` payload will be supplied independently from `index_image` because `index_image` is the pull specification to the manifest list (each manifest being an image for a specific architecture), which gets generated after the images for each architecture are built.

This leads me to believe that we should associate architectures directly with the request instead to avoid needing a results backend.

Reply:
Ah, that makes sense. I agree, we should associate architectures directly with the request instead of index_image. I'll do that as a part of this PR